File: src/main/java/xyz/phanta/cbtweaker/util/TickModulator.java
package xyz.phanta.cbtweaker.util;

public class TickModulator {

    private int interval = 0;
    private int tick = 0;

    public boolean tick() {
        if (interval <= 0) {
            return false;
        } else if (interval == 1) {
            return true;
        } else if (++tick >= interval) {
            tick = 0;
            return true;
        } else {
            return false;
        }
    }

    public void sleep() {
        setInterval(0);
    }

    public void setInterval(int newInterval) {
        interval = newInterval;
        tick = 0;
    }

    public int getInterval() {
        return interval;
    }

    public void increaseInterval(int offset) {
        interval += offset;
    }

    public void increaseIntervalUntil(int offset, int upperBound) {
        interval = Math.min(interval + offset, upperBound);
    }

    public void decreaseInterval(int offset) {
        interval -= offset;
    }

    public void decreaseIntervalUntil(int offset, int lowerBound) {
        interval = Math.max(interval - offset, lowerBound);
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/util/machine/TickModulator.kt
package st.evening.mc.cbtweaker.util.machine

class TickModulator(interval: Int = 0) {
    var interval: Int = interval
        set(value) {
            field = value
            tick = 0
        }
    private var tick: Int = 0

    constructor(active: Boolean) : this (if (active) 1 else 0)

    fun tick(): Boolean {
        if (interval <= 0) {
            return false
        } else if (interval == 1) {
            return true
        } else if (++tick >= interval) {
            tick = 0
            return true
        } else {
            return false
        }
    }

    fun sleep() {
        interval = 0
    }

    fun increaseInterval(offset: Int) {
        interval += offset
    }

    fun increaseIntervalUntil(offset: Int, upperBound: Int) {
        interval = (interval + offset).coerceAtMost(upperBound)
    }

    fun decreaseInterval(offset: Int) {
        interval -= offset
    }

    fun decreaseIntervalUntil(offset: Int, lowerBound: Int) {
        interval = (interval - offset).coerceAtLeast(lowerBound)
    }
}
